lib/auth/grpcserver.go

```diff
@@ -1527,6 +1529,141 @@ func (g *GRPCServer) GetMFADevices(ctx context.Context, req *proto.GetMFADevices
 	}, nil
 }
 
+func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_GenerateUserSingleUseCertsServer) error {
+	actx, err := g.authenticate(stream.Context())
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+
+	// The RPC is streaming both ways and the message sequence is:
+	// (-> means client-to-server, <- means server-to-client)
+	//
+	// 1. -> Init
+	// 2. <- MFAChallenge
+	// 3. -> MFAResponse
+	// 4. <- Certs
+
+	// 1. receive client Init
+	req, err := stream.Recv()
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	initReq := req.GetInit()
+	if initReq == nil {
+		return trail.ToGRPC(trace.BadParameter("expected UserCertsRequest, got %T", req.Request))
+	}
+	if err := validateUserSingleUseCertRequest(initReq, g.AuthServer.GetClock()); err != nil {
+		return trail.ToGRPC(err)
+	}
+
+	// 2. send MFAChallenge
+	// 3. receive and validate MFAResponse
+	if err := userSingleUseCertsAuthChallenge(actx, stream); err != nil {
+		return trail.ToGRPC(err)
+	}
+
+	// Generate the cert.
+	respCert, err := userSingleUseCertsGenerate(stream.Context(), actx, *initReq)
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+
+	// 4. send Certs
+	if err := stream.Send(&proto.UserSingleUseCertsResponse{
+		Response: &proto.UserSingleUseCertsResponse_Cert{Cert: respCert},
+	}); err != nil {
+		return trail.ToGRPC(err)
+	}
+	return nil
+}
+
+func validateUserSingleUseCertRequest(req *proto.UserCertsRequest, clock clockwork.Clock) error {
```

Comment:
Nit: Would it make sense to implement `Check(clock clockwork.Clock)` method on `UserCertsRequest` like we usually do?

Reply:
@Joerger what's the latest advice on where we put validation?
I remember you were trying to remove as much as possible out of `api/`